internal/fanoutconsumer/logs.go

```diff
@@ -59,17 +45,30 @@ func (lsc *logsConsumer) Capabilities() consumer.Capabilities {
 // ConsumeLogs exports the plog.Logs to all consumers wrapped by the current one.
 func (lsc *logsConsumer) ConsumeLogs(ctx context.Context, ld plog.Logs) error {
 	var errs error
-	// Initially pass to clone exporter to avoid the case where the optimization of sending
-	// the incoming data to a mutating consumer is used that may change the incoming data before
-	// cloning.
-	for _, lc := range lsc.clone {
-		clonedLogs := plog.NewLogs()
-		ld.CopyTo(clonedLogs)
-		errs = multierr.Append(errs, lc.ConsumeLogs(ctx, clonedLogs))
+
+	// Clone the data before sending to mutable consumers.
+	// The only exception is the last consumer which is allowed to mutate the data only if there are no
+	// other non-mutating consumers and the data is mutable. Never share the same data between
+	// a mutating and a non-mutating consumer since the non-mutating consumer may process
+	// data async and the mutating consumer may change the data before that.
+	for i, lc := range lsc.mutable {
+		if i < len(lsc.mutable)-1 || ld.IsReadOnly() || len(lsc.readonly) > 0 {
+			clonedLogs := plog.NewLogs()
+			ld.CopyTo(clonedLogs)
+			errs = multierr.Append(errs, lc.ConsumeLogs(ctx, clonedLogs))
+		} else {
+			errs = multierr.Append(errs, lc.ConsumeLogs(ctx, ld))
+		}
 	}
-	for _, lc := range lsc.pass {
+
+	// Mark the data as read-only if it was sent to more than one read-only consumer.
```

Comment:
// Mark the data as read-only if it will be sent to more than one read-only consumer.